### PR TITLE
Batch support for deterministic image ops

### PR DESCRIPTION
**Pull request by JoshVarty**
Based on #14854 and working on #8926

Previously I had implemented batch support for a number of image ops. However performance concerns were raised and the changes were reverted.

I have re-implemented the changes for:
 - `flip_left_right`
- `flip_up_down`
- `transpose_image`
- `rot90`

I ran performance tests from https://github.com/tensorflow/tensorflow/pull/15348 with:

`bazel run -c opt //tensorflow/python:image_ops_test -- --benchmarks=FlipImageBenchmark`

| Operation | Before (μs) | After (μs) |
| --- | --- | --- |
| benchmarkFlipLeftRight_299_299_3_/cpu:0_1 | 274.49  | 264.26 |
| benchmarkFlipLeftRight_299_299_3_/cpu:0__all | 292.76 |  266.10 |
| benchmarkFlipLeftRight_299_299_3___all | 273.80 |  265.71 |
| *benchmarkRandomFlipLeftRight_299_299_3_/cpu:0_1 | 242.58  | 241.89 |
| *benchmarkRandomFlipLeftRight_299_299_3_/cpu:0__all | 245.27 | 239.88 |
| *benchmarkRandomFlipLeftRight_299_299_3___all | 252.71 | 241.20 |

*\* There were no changes made to `RandomFlipLeftRight in this PR*

Let me know if you would like me to add more performance tests for the other methods. I don't think there should be any performance impact, but I'm happy to add more if you'd like.

### CODE REVIEW

**Comment by JoshVarty** [comment]
/cc @martinwicke @jhseu 

I would still like to include support for the random flip ops as well, but I'll do that in another PR if this looks acceptable.

**Comment by rmlarsen** [comment]
@martinwicke could you take a look, please?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
I'm sorry this took me so long. Could you fix the conflicts? I think it's just that proper name_scopes were added to all functions in this module, which is nice and should be easy to merge.

File: tensorflow/python/ops/image_ops_impl.py [hunk 1]
```diff
@@ -292,108 +312,183 @@ def random_flip_left_right(image, seed=None):
...
     k.get_shape().assert_has_rank(0)
     k = math_ops.mod(k, 4)
 
-    def _rot90():
-      return array_ops.transpose(array_ops.reverse_v2(image, [1]), [1, 0, 2])
+    shape = image.get_shape()
+    if shape.ndims == 3 or shape.ndims is None:
+      return _rot90_3D(image, k, scope)
+    elif shape.ndims == 4:
+      return _rot90_4D(image, k, scope)
```
Comment (JoshVarty) [inline tensorflow/python/ops/image_ops_impl.py]: Introducing two methods for `_rot90_3D` and `_rot90_4D` is unfortunate. I'd definitely welcome a better suggestion. The other alternative was to stagger `if shape.ndims == 3` checks throughout the method which didn't really help readability either.

**Comment by JoshVarty** [comment]
@martinwicke I've merged the commits and I think I've done so properly. All the tests pass and there are still no performance regressions so it's probably ready for a code review.

I also introduced `_AssertAtLeast3DImage()` in the style of `_Assert3DImage()` which had been added since I originally submitted this PR.

File: tensorflow/python/ops/image_ops_impl.py (hunk 1 above)
Comment (martinwicke) [reply tensorflow/python/ops/image_ops_impl.py]: I think this is fine.

**Comment by martinwicke** [comment]
Sanity complains about indentation: 
```
FAIL: Found 5 non-whitelited pylint errors:
tensorflow/python/ops/image_ops_impl.py:422: [C0330(bad-continuation), ] Wrong continued indentation (add 2 spaces).

tensorflow/python/ops/image_ops_impl.py:428: [C0330(bad-continuation), ] Wrong continued indentation (add 3 spaces).

tensorflow/python/ops/image_ops_impl.py:437: [C0301(line-too-long), ] Line too long (82/80)

tensorflow/python/ops/image_ops_impl.py:458: [C0330(bad-continuation), ] Wrong continued indentation (add 3 spaces).

tensorflow/python/ops/image_ops_test.py:1125: [C0330(bad-continuation), ] Wrong continued indentation (add 22 spaces).
```

**Comment by JoshVarty** [comment]
Sorry for the delay fixing indentation. I forgot to run the sanity checks locally. :(

Thanks for taking care of it!

**Pull request merged**